internal/system/agent/services_test.go

```diff
@@ -91,164 +102,126 @@ func TestProcessResponse(t *testing.T) {
+	reset()
+	msmMock := &servicesMock.ServiceStarter{}
 
 	action := "start"
-	services := []string{"edgex-config-seed", "edgex-support-logging", "edgex-core-metadata", "edgex-support-notifications", "edgex-core-data", "edgex-core-command", "edgex-export-client", "edgex-export-distro"}
-	params := []string{"graceful"}
+	services := []string{"edgex-core-data", "edgex-core-command"}
 	LoggingClient = logger.NewMockClient()
 
-	res := InvokeOperation(action, services, params)
-	if res != true {
-		t.Errorf("TestInvokeOperationStartKnownService() failed.")
-		return
-	}
-}
-
-func TestInvokeOperationRestart(t *testing.T) {
+	msmMock.On("Start", mock.AnythingOfType("string")).Return(nil)
 
-	action := "restart"
-	services := []string{"edgex-config-seed", "edgex-support-logging", "edgex-core-metadata", "edgex-support-notifications", "edgex-core-data", "edgex-core-command", "edgex-export-client", "edgex-export-distro"}
-	params := []string{"graceful"}
-	LoggingClient = logger.NewMockClient()
+	executorClient = msmMock
 
-	res := InvokeOperation(action, services, params)
-	if res != true {
-		t.Errorf("TestInvokeOperationRestart() failed.")
+	res := InvokeOperation(action, services)
+	if res != nil {
+		t.Errorf("Unexpected error on invoking InvokeOperation() with the START action")
 		return
 	}
 }
 
-func TestInvokeOperationStop(t *testing.T) {
+func TestInvokeOperationStartOperationNotSupported(t *testing.T) {
 
-	action := "stop"
-	services := []string{"edgex-config-seed", "edgex-support-logging", "edgex-core-metadata", "edgex-support-notifications", "edgex-core-data", "edgex-core-command", "edgex-export-client", "edgex-export-distro"}
-	params := []string{"graceful"}
-	LoggingClient = logger.NewMockClient()
-
-	res := InvokeOperation(action, services, params)
-	if res != true {
-		t.Errorf("TestInvokeOperationStop() failed.")
-		return
-	}
-}
-
-func TestInvokeOperationStartUnknownService(t *testing.T) {
+	reset()
+	msmMock := &servicesMock.ServiceStarter{}
 
 	action := "start"
-	services := []string{"foo-bar"}
-	params := []string{"graceful"}
+	services := []string{"edgex-core-data", "edgex-core-command"}
 	LoggingClient = logger.NewMockClient()
 
-	res := InvokeOperation(action, services, params)
-	if res != true {
-		t.Errorf("TestInvokeOperationStartUnknownService() failed.")
-		return
-	}
-}
-
-func TestInvokeOperationRestartUnknownService(t *testing.T) {
+	msmMock.On("Start", mock.AnythingOfType("string")).Return(nil)
 
-	action := "restart"
-	services := []string{"foo-bar"}
-	params := []string{"graceful"}
-	LoggingClient = logger.NewMockClient()
+	executorClient = "operation not supported with specified executor"
 
-	res := InvokeOperation(action, services, params)
-	if res != true {
-		t.Errorf("TestInvokeOperationRestartUnknownService() failed.")
+	res := InvokeOperation(action, services)
+	if res != nil {
+		t.Log("Expected error on invoking InvokeOperation() with the START action")
 		return
 	}
 }
```

Comment:
You might consider using a table-driven testing methodology. For example, I've combined `TestInvokeOperationStart` and `TestInvokeOperationStartOperationNotSupported` into the logic below. With a little thought, this methodology could be extended to facilitate all of the tests in this PR. See what you think.
```
func TestStartOperation(t *testing.T) {
	reset()
	LoggingClient = logger.NewMockClient()
	serviceList := []string{clients.CoreDataServiceKey, clients.CoreCommandServiceKey}
	mockStarter := &servicesMock.ServiceStarter{}
	mockStarter.On("Start", mock.AnythingOfType("string")).Return(nil)

	tests := []struct {
		name		string
		starter 	interface{}
		services	[]string
		expectError bool
	} {
		{"start services", mockStarter, serviceList, false},
		{"type check failure", "abc", serviceList, true},
	}
	for _, tt := range tests {
		executorClient = tt.starter
		t.Run(tt.name, func(t *testing.T) {
			err := InvokeOperation("start", tt.services)
			if !tt.expectError && err != nil {
				t.Errorf("unexpected error: %v", err)
			}
			if tt.expectError && err == nil {
				t.Errorf("did not receive expected error: %s", tt.name)
			}
		})
	}
}
```

Reply:
@tsconn23 I like this very much, and will follow up 👍 In fact, I came across the notion of `a table-driven testing methodology` in just the past few days, while reading a book recommended by @michaelestrin (It's called `Hands-On Dependency Injection in Go`, and has several excellent ideas.)